### PR TITLE
Make use of service loader mechanism for JsonbComponentInstanceCreator

### LINKED ISSUE #260: Turn JsonbComponentInstanceCreator into a service provider interface
So that an integrator could provide a custom implementation of this interface. `JsonbComponentInstanceCreatorFactory` currently contains some reflection magic to detect the presence of a CDI container. In fact, the presence of the CDI API is enough.

I believe it would be reasonable to define this SPI. [Quarkus](https://github.com/quarkusio/quarkus) integration would definitely make use of it because our DI solution is compatible with CDI but we don't support all "portable extensions related" SPIs (they do not fit into our build time metadata processing approach) and so the invocation of  the `BeanManagerInstanceCreator#getOrCreateComponent()` fails.

### PR DESCRIPTION
**Pull request by mkouba**
- if there is no service provider use the default implementations
- resolves #260

### CODE REVIEW

**Comment by mkouba** [comment]
I'm not sure how to test this feature properly. Nevertheless, I've tried to implement a `JsonbComponentInstanceCreator` service provider in Quarkus and it seems to work.

File: src/main/java/org/eclipse/yasson/internal/JsonbContext.java [hunk 1]
```diff
@@ -130,4 +141,23 @@ public JsonbConfigProperties getConfigProperties() {
...
+        for (JsonbComponentInstanceCreator creator : loader) {
+            creators.add(creator);
+        }
+        if (creators.isEmpty()) {
+            // No service provider found - use the defaults
+            return JsonbComponentInstanceCreatorFactory.getComponentInstanceCreator(instanceCreator);
+        }
+        creators.sort(Comparator.comparingInt(JsonbComponentInstanceCreator::getPriority).reversed());
+        JsonbComponentInstanceCreator creator = creators.get(0);
+        log.finest(Messages.getMessage(MessageKeys.SERVICE_PROVIDER_COMPONENT_INSTANCE_CREATOR, creator.getClass()));
```
Comment (aguibert) [inline src/main/java/org/eclipse/yasson/internal/JsonbContext.java]: using an external message resource isn't necessary for a log.finest statement, just INFO or higher messages and exception messages need that.

File: src/main/java/org/eclipse/yasson/internal/components/JsonbComponentInstanceCreator.java [hunk 2]
```diff
@@ -15,13 +15,21 @@
...
+ * 
+ * <p>
+ * Yasson attempts to load the implementations using {@link java.util.ServiceLoader} first. If there are multiple
+ * implementations found the service provider with the highest priority is used. If there are no implementations found the
+ * {@link JsonbComponentInstanceCreatorFactory} is used to instantiate the default implementation.
+ * </p>
+ * 
  * @author Roman Grigoriadi
  */
 public interface JsonbComponentInstanceCreator extends Closeable {
```
Comment (aguibert) [inline src/main/java/org/eclipse/yasson/internal/components/JsonbComponentInstanceCreator.java]: lets move this interface to a new `org.eclipse.yasson.spi` package and export it in the module-info.java. If we left this interface here, other libraries would not be able to actually implement this interface under the constraints of JPMS.

File: src/main/java/org/eclipse/yasson/internal/JsonbContext.java (hunk 1 above)
Comment (mkouba) [reply src/main/java/org/eclipse/yasson/internal/JsonbContext.java]: Ok. I was just following the usage from [JsonbComponentInstanceCreatorFactory](https://github.com/eclipse-ee4j/yasson/blob/master/src/main/java/org/eclipse/yasson/internal/components/JsonbComponentInstanceCreatorFactory.java#L58).

File: src/main/java/org/eclipse/yasson/internal/components/JsonbComponentInstanceCreator.java (hunk 2 above)
Comment (mkouba) [reply src/main/java/org/eclipse/yasson/internal/components/JsonbComponentInstanceCreator.java]: Ok. To be honest I did not consider JPMS at all.

**Comment by bravehorsie** [comment]
Using this service provider would make application need to depend on Yasson instead of Jsonb API. Does it worth proposing to move `JsonbComponentInstanceCreator` in the Jsonb API? How about adding a switch into JsonbConfig to shutdown CDI completely even when it is discoverable?

**Comment by mkouba** [comment]
> Using this service provider would make application need to depend on Yasson instead of Jsonb API. Does it worth proposing to move JsonbComponentInstanceCreator in the Jsonb API? 

I don't think so. This is an impl-specific integration.

File: src/main/java/module-info.java [hunk 3]
```diff
@@ -19,5 +19,7 @@
     requires static java.desktop;
     
     exports org.eclipse.yasson;
+    exports org.eclipse.yasson.spi;
     provides javax.json.bind.spi.JsonbProvider with org.eclipse.yasson.JsonBindingProvider;
+    uses org.eclipse.yasson.internal.components.JsonbComponentInstanceCreator;
```
[the remark this answers is not in the document]
Comment (mkouba) [reply src/main/java/module-info.java]: Ah, sure. Sorry.

File: src/main/java/org/eclipse/yasson/internal/JsonbContext.java [hunk 4]
```diff
@@ -130,4 +139,23 @@ public JsonbConfigProperties getConfigProperties() {
     public InstanceCreator getInstanceCreator() {
         return instanceCreator;
     }
+
+    private JsonbComponentInstanceCreator initComponentInstanceCreator(InstanceCreator instanceCreator) {
+        ServiceLoader<JsonbComponentInstanceCreator> loader = AccessController
```
Comment (bravehorsie) [inline src/main/java/org/eclipse/yasson/internal/JsonbContext.java]: This also applies to further calls discovering Jsonp, but calling service loader slows down cold start of Jsonb (`JsonbBuilder.create()`). Do we want to add some classloader based cache here in future?

Comment (mkouba) [reply src/main/java/org/eclipse/yasson/internal/JsonbContext.java]: Well, although it sounds like a good idea the caches are often tricky. And if I understand it correctly users are encouraged to resuse `Jsonb` and `JsonbBuilder` instances for optimal performance.

Comment (bravehorsie) [reply src/main/java/org/eclipse/yasson/internal/JsonbContext.java]: Yes Jsonb instance is encouraged to be cached. If done, similar caching is not needed. However there is number of [reappearing similar requests](https://github.com/eclipse-ee4j/jsonp/issues/154) in Jsonp. Using APP classloader may solve it for both even when instances are not cached.

Comment (mkouba) [reply src/main/java/org/eclipse/yasson/internal/JsonbContext.java]: I'm still not convinced that we should cache this but feel free to address this problem in another PR ;-)

Comment (bravehorsie) [reply src/main/java/org/eclipse/yasson/internal/JsonbContext.java]: Sure, I am not asking to fix it here.

**Comment by bravehorsie** [comment]
Looks good to me to. Merging.

**Comment by bravehorsie** [comment]
Thank you @mkouba !

**Comment by mkouba** [comment]
It’s a pleasure. Do you plan to release 1.0.4 in the near future?

**Pull request merged**